File: Striver Important Interview Questions/Day 19 - Binary Trees Part 3/binary-tree-maximum-path-sum.cpp

// https://leetcode.com/problems/binary-tree-maximum-path-sum/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int maxPathSumHelper(TreeNode*root, int &ans){
        if(!root) return 0;
        int leftAns = max(0, maxPathSumHelper(root->left, ans));
        int rightAns = max(0, maxPathSumHelper(root->right, ans));
        ans = max(ans, root->val + leftAns + rightAns);
        return root->val + max(leftAns, rightAns);
    }

    int maxPathSum(TreeNode* root) {
        int ans = INT_MIN;
        if(!root) return ans;
        maxPathSumHelper(root, ans);
        return ans;
    }
};

File: Neetcode-150-js/Trees/binary-tree-maximum-path-sum.js

// https://leetcode.com/problems/binary-tree-maximum-path-sum/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    if(!root) return 0;
    let maxValue = Number.MIN_SAFE_INTEGER;
  
    var check = (root) => {
      if(!root) return 0;
      let leftAns = Math.max(check(root.left), 0);
      let rightAns = Math.max(check(root.right), 0);
      maxValue = Math.max(maxValue, root.val + leftAns + rightAns);
      return root.val + Math.max(leftAns, rightAns);
    } 
  
      check(root);
      return maxValue;
  };